internal/gossip/gossip.go

```diff
@@ -0,0 +1,101 @@
+package gossip
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/facebookgo/startstop"
+	"golang.org/x/sync/errgroup"
+)
+
+// Gossiper is an interface for broadcasting messages to all receivers
+// subscribed to a channel
+type Gossiper interface {
+	// Publish sends a message to all peers listening on the channel
+	Publish(channel string, value []byte) error
+
+	// Subscribe listens for messages on the channel
+	Subscribe(channel string, callback func(data []byte)) error
+
+	startstop.Starter
+	startstop.Stopper
+}
+
+var _ Gossiper = &InMemoryGossip{}
+
+// InMemoryGossip is a Gossiper that uses an in-memory channel
+type InMemoryGossip struct {
+	channel     chan []byte
+	subscribers map[string][]func(data []byte)
+
+	done chan struct{}
+	eg   *errgroup.Group
+}
+
+func (g *InMemoryGossip) Publish(channel string, value []byte) error {
+	msg := message{
+		key:  channel,
+		data: value,
+	}
+
+	select {
+	case <-g.done:
+		return errors.New("gossip has been stopped")
+	case g.channel <- msg.ToBytes():
+	default:
```

Comment:
I don't love the default case here; it could lead to us dropping messages a bit too easily. Can we please give the channel a length (like 10)?